libpod/image/image.go

```diff
@@ -1076,3 +1078,144 @@ func (i *Image) Comment(ctx context.Context, manifestType string) (string, error
+func (i *Image) Tree(ctx context.Context, imageName string) (*bytes.Buffer, error) {
+
+	layerMap, err := i.layerToImageMap(ctx)
+	if err != nil {
+		return nil, err
+	}
+	layerMap, err = i.addImageToLayerInfo(ctx, layerMap)
+
+	repoTag, err := i.MatchRepoTag(imageName)
+	if err != nil {
+		return nil, err
+	}
+	// Get the image top layer and then parse as per parent
+	layer := i.TopLayer()
+	dependencyList := returnDependencyTree(layerMap, repoTag, layer)
+
+	var buffer bytes.Buffer
+	printTree(&buffer, dependencyList)
+
+	return &buffer, err
+}
+
+func printTree(buffer *bytes.Buffer, treeStack []TreeImage) {
+
+	prefixSpace := 0
+	for i := len(treeStack) - 1; i >= 0; i-- {
+		for p := 0; p < prefixSpace; p++ {
+			buffer.WriteString(fmt.Sprintf(" "))
+		}
+		buffer.WriteString(fmt.Sprintf("└─ ID: %s  Tags: %s VirtualSize: %v\n",
+			treeStack[i].ID[:12],
+			treeStack[i].RepoTags,
+			units.HumanSizeWithPrecision(float64(treeStack[i].Size), 4)),
+		)
+
+		prefixSpace++
+	}
+}
+
+func (i *Image) layerToImageMap(ctx context.Context) (map[string]*TreeImage, error) {
+	layerMap := make(map[string]*TreeImage)
+	layers, _ := i.imageruntime.store.Layers()
+	for _, layer := range layers {
+		existingLayer, ok := layerMap[layer.ID]
+		if !ok {
+			layerMap[layer.ID] = &TreeImage{
+				ID:       layer.ID,
+				Size:     layer.CompressedSize,
+				ParentID: layer.Parent,
+			}
+		} else {
+			return nil, fmt.Errorf("duplicate %s layer found in image store", existingLayer.ID)
+		}
+	}
+	return layerMap, nil
+}
+
+func (i *Image) addImageToLayerInfo(ctx context.Context, layerMap map[string]*TreeImage) (map[string]*TreeImage, error) {
+	images, err := i.imageruntime.GetImages()
+	if err != nil {
+		return nil, errors.Wrapf(err, "error getting images from store")
+	}
+
+	for _, image := range images {
+
+		layerid := image.TopLayer()
+		imgLayer, err := image.imageruntime.store.Layer(layerid)
+		if err != nil {
+			return nil, err
+		}
+
+		for imgLayer != nil {
+			existingLayer, ok := layerMap[imgLayer.ID]
+			if !ok {
+				return nil, fmt.Errorf("layer %s not found ", imgLayer.ID)
+			} else {
+				names := image.Names()
+				existingLayer.RepoTags = append(existingLayer.RepoTags, names...)
+			}
+			imgLayer = image.getLayer(imgLayer.Parent)
+		}
+	}
+	return layerMap, nil
+}
+
+func (i *Image) getLayer(layerID string) *storage.Layer {
+	layers, err := i.imageruntime.store.Layers()
+	if err != nil {
+		return nil
+	}
+
+	for _, layer := range layers {
+		if layerID == layer.ID {
+			return &layer
+		}
+	}
+	return nil
+}
+
+func returnDependencyTree(layerMap map[string]*TreeImage, repoTag string, layerID string) []TreeImage {
+	var treeStack []TreeImage
+	previous := TreeImage{}
+	layerInfo, ok := layerMap[layerID]
+	for ok {
+		appended := false
+		// First loop
+		if len(previous.RepoTags) == 0 {
+			previous.ID = layerInfo.ID
+			previous.RepoTags = []string{repoTag}
+		}
+
+		if previous.RepoTags[0] == layerInfo.RepoTags[0] {
```

Comment:
> The way `addImageToLayerInfo` works AFAICS, the first image it encounters is recorded in `RepoTags` of its complete layer sequence, i.e. `previous.RepoTags[0] == layerInfo.RepoTags[0]` will almost always be true, unless `addImageToLayerInfo` just happens to first encounter parent images and then derived images.

Reply:
> (Now that layerMap gives O(1) layer access, I’d expect an O(layers) pass to build lists of, or at least count, children of each layer, and use that to determine the logical shape of the graph.)

That was my original idea, to use each `Repo Name` to show branch. I dropped for this PR due to its complexity. The link `(Parent <-> child )` is bottom to top, which makes it difficult to do it in one pass.

Reply:
Well, reversing the links is just a single O(layers) pass:
```
for layer in layers:
   if layer.Parent != "":
     parent := layers[layer.Parent]
     parent.children.append(layer)
```

(or maybe just count the children instead of creating lists, if the lists are not necessary), and then this code to skip irrelevant layers can use a more precise logic for layers that are “irrelevant” (maybe something like “has exactly one child and no tag”).

---

Either way, due to the way `RepoTags[0]` is an unpredictable choice of all the possible names, the check as is AFAICS causes this function to behave similarly unpredictably when choosing which layers to omit.  So, either implement some other, consistent, logic for omitting layers, or, to simplify, for now, don’t omit any.